### ISSUE TITLE
Feature request: XPRMNTL to not allow app.js to add new experiments

### ISSUE THREAD

**Issue by ghost** [XPRMNTL/xpr-dashboard#1]
Currently, the only way to create a new experiment on XPRMNTL is to add it to your `app.js` file's `expList` object and restart the server. When doing localhost modifying of the `expList`, you can push a malformed object to XPRMNTL. This caused Tree to go down for an hour because there was a bad experiment. 

What I am proposing is that XPRMNTL no longer allow apps to add new experiments through the `app.js` file. Instead, the XPRMNTL dashboard would allow you to add new experiments to an app, and then apps would read that list through the normal means. This would prevent the malformed object problem and the problem of the `expList` being out of sync with XPRMNTL (since it's hard to maintain the default list when you make a change on XPRMNTL).

**Comment by jshcrowthe** [comment]
I believe that the idea behind having the `app.js` be the master experiment list was in the case of a request failure the app still has an accessible version of the experiment that it can default to (@dncrews correct me if I'm wrong). If we were to allow adding experiments via the dashboard, _in the event of a request error,_ there would be no local backup available to the app. Because of this all experiments created via the dashboard (and not in the `app.js`) would all be hidden regardless. So I would maintain that the `app.js` should stay the source of experiments.

All that being said I would totally support adding some verification so that no malformed objects are making it into the experiment list. That is something that I am surprised doesn't already exist.

**Comment by ghost** [comment]
Ya, we've been discussing the failure problem as the 2nd half to this request, but since it wasn't part of XPRMNTL but instead part of our app, I didn't want to include it in the feature request. 

Essentially we are trying to come up with a means for our app to write some sort of fallback experiment list to each environment when it requests the experiments (I hear @dncrews wanted to do it this way first anyway). Then when the request does fail, it would just read from that list (which would always be up-to-date defaults).

We're exploring different options, but having `app.js` be the source is very dangerous when the request does fail since apps aren't maintaining the list of defaults at all (it's too separated at the moment since `app.js` is just used to create but not maintain experiment defaults). The day the request does fail, we'll have some big problems anyway.

**Comment by ghost** [comment]
Another problem with `app.js` being the source of experiments is that you cannot have separate defaults for each environment, the defaults must be prod like.

**Comment by dncrews** [comment]
That's correct. In the first iteration of this, the expectation is that in the case of failure, only **released** features would remain enabled, which would have the default `on`. That didn't matter very much since "failure" meant that there was a deploy/restart happening successfully while the service was "down", which wasn't likely.

**Comment by dncrews** [comment]
I definitely agree that the low-hanging fruit here would be validation, which (honestly) I too am surprised didn't exist already.

**Comment by dncrews** [comment]
If it was important enough, I could definitely see building a plugin for environment-specific defaults.

**Comment by waaronking** [comment]
> All that being said I would totally support adding some verification so that no malformed objects are making it into the experiment list.

+1 for the verification. 

I haven't experienced any issues of the app.js list and XPRMNTL being out of sync (since I always change my local before going up and changing XPRMNTL). I personally believe that is a training issue, but I would chime in with this thought:

We had two issues today where people were developing on localhost with "new experiments." Since they were local, they thought, "No big deal if I add this new experiment with the state of TRUE while I test things." Then, before making a PR and pushing, they set the experiment to FALSE and made their PR. As things ended up, the experiments went LIVE as soon as the PR got merged (because that was the default state when they first made the experiment--while they were testing locally).

On the one hand, this could definitely be a training issue, but on the other--it could be nice if there was a feature that allowed all new experiments to default to FALSE for PROD. That way playing in the local playground won't accidentally sabotage production.

Another solution would be to "approve" experiments on XPRMNTL before they can turn on. This could make the most sense... maybe I should make a PR ;)

**Comment by ghost** [comment]
There is also the issue that any branches/forks of the app can also push out any experiment to the dashboard with localhost.

**Comment by dncrews** [comment]
@waaronking, something could be added that makes sure new experiments are always OFF by default... The approval concept is an interesting one...

**Comment by waaronking** [comment]
Ok, I've made two pull requests that should allow for a default OFF state. I've tested it by deploying my own `feature` server: http://fs-feature2.herokuapp.com/ It works as expected. Let me know your thoughts
- https://github.com/XPRMNTL/feature-client.js/pull/1
- https://github.com/XPRMNTL/feature/pull/2

**Comment by dncrews** [comment]
So I'm thinking: should this be something at the app-level, or a global configuration? I would think it should be a setting in the dashboard. We don't have any kind of Admin panel, but an env variable might be a good polyfill until then. Anyone have thoughts?

**Comment by waaronking** [comment]
I could potentially see the use case for both app and global configs. Both features could co-exist. An admin panel makes a lot of sense--user rights would need to come at the same time.

**Comment by spacerockzero** [comment]
☎️ Long-time listener, first-time caller...

I'm re-visiting this now for a different reason.

We are having a lot of deleted experiments pop back up on the dashboard when people start up older versions of shared code in their apps. (Zombie Experiment problem)
It's been frustrating enough with the number of our devs using previous pins of shared repos and seeing this, that we need to have an option to switch the experiment addition from app broadcast, to add explicitly via XPRMNTL dashboard. This is the pattern that LaunchDarkly uses. I figure it could be a welcome addition to XPRMNTL.

I can make a PR with such functionality. We have a story running this iteration to solve this problem.
